UPGRADE.md

```diff
@@ -2798,3 +2798,96 @@ import { BooleanInput } from 'react-admin';
 +   onChange={e => { console.log(e.target.checked); }}
 />
 ```
+
+## `<ReferenceInput>` and `<ReferenceArrayInput>` Now Provide A `ChoicesContext`
+
+`<ReferenceInput>` and `<ReferenceArrayInput>` used to inject props to their children. They now provide a `ChoicesContext` with support for sorting, paginating and filtering through methods and properties that match those of the `ListContext`. All their methods now have the same signature. As a result, several changes were made to both components:
+
+### `<ReferenceInput>` No Longer Accepts The `filterToQuery` Prop
+
+It is now the responsibility of the child input to call the `setFilters` function provided through the `ChoicesContext` with the expected filter. If you used the `filterToQuery` prop to convert an `<AutocompleteInput>` search, you'll have to move the `filterToQuery` prop on the `<AutocompleteInput>` itself:
+
+```diff
+const UserListFilter = [
+    <ReferenceInput
+        source="email"
+-        filterToQuery={search => ({ email: search })}
+    >
+-        <AutocompleteInput />
++        <AutocompleteInput filterToQuery={search => ({ email: search })} />
+    </ReferenceInput>
+]
+```
+
+### `<ReferenceArrayInput>` No Longer Provides a `ListContext`
+
+As the `ChoicesContext` now provide an API very similar to the `ListContext`, it no longer sets up a `ListContext`. If you used this to display a `<Datagrid>`, we will soon provide the `<DatagridInput>` for this purpose.
```

Comment:
😄 I like this futuristic comment

Reply:
There's a WIP version in this PR already!